MSAL/src-internal/ios/MSALLegacySharedAccountsProvider.m

```diff
@@ -0,0 +1,512 @@
+    {
+        MSALAccountEnumerationParameters *parameters = [MSALLegacySharedAccountFactory parametersForAccount:account
+                                                                                    tenantProfileIdentifier:account.accountClaims[@"oid"]];
+        
+        if (!parameters)
+        {
+            NSError *parameterError = MSIDCreateError(MSIDErrorDomain, MSIDErrorInternal, @"Unable to create parameters for the account", nil, nil, nil, nil, nil);
+            [self fillAndLogError:error withError:parameterError logLine:@"Failed to create parameters for the account"];
+            return nil;
+        }
+        
+        return [self accountsFromJsonObject:jsonDictionary withParameters:parameters error:error];
+    }
+    
+    NSMutableArray *allAccounts = [NSMutableArray new];
+    
+    for (MSALTenantProfile *tenantProfile in tenantProfiles)
+    {
+        MSALAccountEnumerationParameters *parameters = [MSALLegacySharedAccountFactory parametersForAccount:account
+                                                                                    tenantProfileIdentifier:tenantProfile.identifier];
+        
+        if (!parameters)
+        {
+            NSError *parameterError = MSIDCreateError(MSIDErrorDomain, MSIDErrorInternal, @"Unable to create parameters for the account", nil, nil, nil, nil, nil);
+            [self fillAndLogError:error withError:parameterError logLine:@"Failed to create parameters for the account"];
+            return nil;
+        }
+        
+        NSArray<MSALLegacySharedAccount *> *accounts = [self accountsFromJsonObject:jsonDictionary withParameters:parameters error:error];
+        
+        if (!accounts)
+        {
+            return nil;
+        }
+        
+        [allAccounts addObjectsFromArray:accounts];
+    }
+    
+    return allAccounts;
+}
+
+#pragma mark - Write
+
+- (BOOL)updateAccount:(id<MSALAccount>)account
+        idTokenClaims:(NSDictionary *)idTokenClaims
+       tenantProfiles:(NSArray<MSALTenantProfile *> *)tenantProfiles
+            operation:(MSALLegacySharedAccountWriteOperation)operation
+                error:(NSError **)error
+{
+    __block BOOL result = YES;
+    __block NSError *updateError = nil;
+    
+    dispatch_barrier_sync(self.synchronizationQueue, ^{
+        result = [self updateAccountImpl:account
+                           idTokenClaims:idTokenClaims
+                          tenantProfiles:tenantProfiles
+                               operation:operation
+                                   error:&updateError];
+    });
+    
+    if (error && updateError)
+    {
+        *error = updateError;
+    }
+    
+    return result;
+}
+
+- (BOOL)updateAccountImpl:(id<MSALAccount>)account
+            idTokenClaims:(NSDictionary *)idTokenClaims
+           tenantProfiles:(NSArray<MSALTenantProfile *> *)tenantProfiles
+                operation:(MSALLegacySharedAccountWriteOperation)operation
+                    error:(NSError **)error
```

Comment:
Can we refactor this function. It seems kind of long.

Reply:
any suggestions for how? It tries to combine similar functionality and save a lot of otherwise similar code between update and remove operations

Reply:
I guess this is a tradeoff between being able to follow the code easily and code reuse/structure. I'm open to splitting it up into smaller functions, but personally don't see a huge need here.